pkg/discovery/systemd.go

```diff
@@ -50,111 +45,66 @@ func (c *SystemdConfig) NewDiscoverer(d DiscovererOptions) (Discoverer, error) {
-
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-ticker.C:
-		}
-
-		units, err := c.units(ctx)
-		if err != nil {
-			return fmt.Errorf("failed to list units: %w", err)
-		}
-		newSourceList := map[string]struct{}{}
-		var targetGroups []*Group
-		for _, unit := range units {
-			labelset := model.LabelSet{
-				"systemd_unit": model.LabelValue(unit),
-			}
-			pids, err := c.pids(ctx, unit)
-			if err != nil {
-				level.Debug(c.logger).Log("msg", "failed to get PIDs for unit", "unit", unit, "err", err)
-				continue
-			}
-
-			newSourceList[unit] = struct{}{}
-
-			targetGroups = append(targetGroups, &Group{
-				Targets: []model.LabelSet{{}},
-				Labels:  labelset,
-				Source:  unit,
-				PIDs:    pids,
-			})
-		}
-
-		// Add empty groups for targets which have been removed since the previous run.
-		for unit := range c.oldSourceList {
-			if _, ok := newSourceList[unit]; !ok {
-				targetGroups = append(targetGroups, &Group{Source: unit})
-			}
-		}
-		c.oldSourceList = newSourceList
-
-		up <- targetGroups
-	}
-}
-
-func (c *SystemdDiscoverer) units(ctx context.Context) ([]string, error) {
-	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
-	defer cancel()
-
-	cmd := exec.CommandContext(ctx, "systemctl", "list-units", "--type=service", "--state=running")
-	stdout, err := cmd.StdoutPipe()
+	conn, err := systemd.NewWithContext(ctx)
 	if err != nil {
-		return nil, err
-	}
-	if err := cmd.Start(); err != nil {
-		return nil, err
+		level.Error(c.logger).Log("msg", "failed to connect to systemd D-Bus API", "err", err)
 	}
+	defer conn.Close()
 
-	units := parseUnitList(stdout)
-	if err := cmd.Wait(); err != nil {
-		return nil, err
+	isSubStateChanged := func(u1, u2 *systemd.UnitStatus) bool {
+		return u1.SubState != u2.SubState
 	}
-	return units, nil
-}
 
-func parseUnitList(stdout io.Reader) []string {
-	units := []string{}
-	s := bufio.NewScanner(stdout)
-	for s.Scan() {
-		line := strings.TrimSpace(s.Text())
-		if strings.HasPrefix(line, "UNIT") ||
-			strings.HasPrefix(line, "LOAD") ||
-			strings.HasPrefix(line, "ACTIVE") ||
-			strings.HasPrefix(line, "SUB") {
-			continue
-		}
-		fields := strings.Fields(line)
-		if len(fields) < 1 {
-			continue
-		}
-		units = append(units, fields[0])
+	isService := func(name string) bool {
+		return !strings.HasSuffix(name, ".service")
 	}
-	if len(units) > 0 {
-		units = units[:len(units)-1]
-	}
-	return units
-}
 
-func (c *SystemdDiscoverer) pids(ctx context.Context, unit string) ([]int, error) {
-	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
-	defer cancel()
+	updateCh, errCh := conn.SubscribeUnitsCustom(time.Second, 0, isSubStateChanged, isService)
 
-	cmd := exec.CommandContext(ctx, "systemctl", "show", "--property", "MainPID", "--value", unit)
-	output, err := cmd.Output()
-	if err != nil {
-		return nil, err
-	}
-
-	pid, err := strconv.Atoi(strings.TrimSpace(string(output)))
-	if err != nil {
-		return nil, err
+	for {
```

Comment:
Why do we need the outer loop? It seems to me we could fold into one.
What I am missing?

Reply:
Nothing, I hadn't noticed that, thank you for catching it!